File: pt-1/527.word-abbreviation.go
package leetcode

//
// @lc app=leetcode id=527 lang=golang
//
// [527] Word Abbreviation
//
// Notes
// Start every word with prefix length 1. Repeatedly group words by their current
// abbreviation; whenever a group conflicts, increase the prefix length for all
// words in that group. When no group conflicts, abbreviations are unique and
// still as short as this incremental process permits. Time: O(rounds*n*L).
// Space: O(nL).
//
// @lc code=start

import "strconv"

func WordsAbbreviation527(words []string) []string {
	prefixLengths := make([]int, len(words))
	for i := range prefixLengths {
		prefixLengths[i] = 1
	}

	for {
		groups := map[string][]int{}
		for index, word := range words {
			abbr := abbreviate527(word, prefixLengths[index])
			groups[abbr] = append(groups[abbr], index)
		}

		hasConflict := false
		for _, indices := range groups {
			if len(indices) <= 1 {
				continue
			}
			hasConflict = true
			for _, index := range indices {
				prefixLengths[index]++
			}
		}
		if !hasConflict {
			break
		}
	}

	answer := make([]string, len(words))
	for index, word := range words {
		answer[index] = abbreviate527(word, prefixLengths[index])
	}
	return answer
}

func abbreviate527(word string, prefixLength int) string {
	omitted := len(word) - prefixLength - 1
	abbreviation := word[:prefixLength] + strconv.Itoa(omitted) + word[len(word)-1:]
	if len(abbreviation) < len(word) {
		return abbreviation
	}
	return word
}

// @lc code=end

File: pt-1/527.word-abbreviation.java
/*
 * @lc app=leetcode id=527 lang=java
 *
 * [527] Word Abbreviation
 *
 * Maintain one prefix length per word. Recompute abbreviations, group equal
 * abbreviations, and increase prefix length only for conflicting groups until
 * every abbreviation is unique or no shorter than the original.
 *
 * Java note: HashMap<String,List<Integer>> is used for conflict grouping.
 *
 * Time: O(rounds * n * word length). Space: O(n).
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// @lc code=start
class Solution {
    public List<String> wordsAbbreviation(List<String> words) {
        int n = words.size();
        int[] prefixLengths = new int[n];
        for (int i = 0; i < n; i++) {
            prefixLengths[i] = 1;
        }

        while (true) {
            Map<String, List<Integer>> groups = new HashMap<>();
            for (int i = 0; i < n; i++) {
                groups.computeIfAbsent(abbreviate(words.get(i), prefixLengths[i]), unused -> new ArrayList<>()).add(i);
            }

            boolean hasConflict = false;
            for (List<Integer> indices : groups.values()) {
                if (indices.size() <= 1) {
                    continue;
                }
                hasConflict = true;
                for (int index : indices) {
                    prefixLengths[index]++;
                }
            }
            if (!hasConflict) {
                break;
            }
        }

        List<String> answer = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            answer.add(abbreviate(words.get(i), prefixLengths[i]));
        }
        return answer;
    }

    private String abbreviate(String word, int prefixLength) {
        int omitted = word.length() - prefixLength - 1;
        String abbreviation = word.substring(0, prefixLength) + omitted + word.charAt(word.length() - 1);
        return abbreviation.length() < word.length() ? abbreviation : word;
    }
}
// @lc code=end
